codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SemanticVersion.java

```diff
@@ -0,0 +1,327 @@
+ */
+public final class SemanticVersion {
+    private final String prefix;
+    private final int major;
+    private final int minor;
+    private final int patch;
+    private final String extra;
+
+    private SemanticVersion(Builder builder) {
+        prefix = builder.prefix;
+        major = builder.major;
+        minor = builder.minor;
+        patch = builder.patch;
+        extra = builder.extra;
+    }
+
+    /**
+     * The semantic version prefix present before the major version.
+     *
+     * @return the optional prefix
+     */
+    public Optional<String> getPrefix() {
+        return Optional.ofNullable(prefix);
+    }
+
+    /**
+     * The major version number.
+     *
+     * @return the major version
+     */
+    public int getMajor() {
+        return major;
+    }
+
+    /**
+     * The minor version number.
+     *
+     * @return the minor version
+     */
+    public int getMinor() {
+        return minor;
+    }
+
+    /**
+     * The patch version number.
+     *
+     * @return the patch version
+     */
+    public int getPatch() {
+        return patch;
+    }
+
+    /**
+     * The optional extra string containing the pre-release and build information.
+     *
+     * @return the optional extra string
+     */
+    public Optional<String> getExtra() {
+        return Optional.ofNullable(extra);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder();
+
+        if (getPrefix().isPresent()) {
+            builder.append(getPrefix().get());
+        }
+
+        builder.append(getMajor());
+        builder.append('.');
+        builder.append(getMinor());
+        builder.append('.');
+        builder.append(getPatch());
+        if (getExtra().isPresent()) {
+            builder.append(getExtra().get());
+        }
+
+        return builder.toString();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        SemanticVersion that = (SemanticVersion) o;
+        return getMajor() == that.getMajor()
+                && getMinor() == that.getMinor()
+                && getPatch() == that.getPatch()
+                && getPrefix().equals(that.getPrefix())
+                && getExtra().equals(that.getExtra());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getPrefix(), getMajor(), getMinor(), getPatch(), getExtra());
+    }
+
+    /**
+     * Parse a semantic version string into a {@link SemanticVersion}.
+     *
+     * @param version the semantic version string
+     * @return the SemanticVersion representing the parsed value
+     */
+    public static SemanticVersion parseVersion(String version) {
+        String[] components = version.split("\\.", 3);
+
+        if (components.length != 3) {
+            throw newInvalidSemanticVersion();
```

Comment:
Nit: can we return the version value in the error?

Reply:
Also which dependency if possible, but that might be not available at this scope.

Reply:
Yeah unfortunately not known in this scope.